docs/design/storagenode-graceful-exit-protocol.md

```diff
@@ -0,0 +1,215 @@
+
+### Sketch
+
+```
+### storagenode ###
+
+for {
+    stream := dial satellite
+
+    for {
+        msg := stream.Recv(&msg)
+        if msg is NotReady {
+            go sleep a bit
+            and retry later
+        }
+
+        if msg is Completed {
+            update database about completion
+            delete all pieces
+            exit
+        }
+
+        if msg is TransferConfirmed {
+            update the progress table
+            delete the piece from disk
+            exit
+        }
+
+        if msg is TransferPiece {
+            // transfer multiple pieces in parallel, configurable
+            limiter.Go(func(){
+                result := try transfer msg.PieceID
+                stream.Send(result)
+                update progress table
+            })
+        }
+    }
+}
+
+### satellite ###
+
+if !initiated {
+    then add node to the gracefully exiting list
+    send NotReady
+    return
+}
+
+if !pieces collected {
+    send NotReady
+    return
+}
+
+inprogress pieces
+more pieces := true
+
+when storage node prematurely exits
+    we'll mark every inprogress as failed
+
+go func() {
+    for {
+        ensure we have only up to N inprogress at the same time
+        
+        list transferred piece that is not in progress
+        if no pieces {
+            morepieces = false
+            break
+        }
+
+        stream.Send TransferPiece
+    }
+
+    more pieces = false
+}()
+
+for more pieces && len(inprogress) > 0 {
+    response := stream.Recv
+
+    verify that response has proper signatures and things
+    update metainfo database with the new storage node
+    delete from inprogress
+
+    stream.Send TransferConfirmed
+}
+
+stream.Send Completed with receipt
+```
+
+
+
+``` proto
+service GracefulExit {
+    rpc Process(stream StorageNodeMessage) returns (stream SatelliteMessage)
+}
+
+message StorageNodeMessage {
+    oneof Message {
+        message TransferSucceeded {
+            AddressedOrderLimit addressed_order_limit;
+            PieceHash original_piece_hash;
+            PieceHash replacement_piece_hash;
+        }
+        message TransferFailed {
+            enum Error {
+                NOT_FOUND = 0;
+                STORAGE_NODE_UNAVAILABLE = 1;
+                UNKNOWN = 2;
+            }
+            Error error = 0;
+        }
+    }
+}
+
+message SatelliteMessage {
+    oneof Message {
+        message NotReady {} // this could be a grpc error rather than a message
+
+        message TransferPiece {
+            bytes piece_id; // the current piece-id
+            bytes private_key;
+            // addressed_order_limit contains the new piece id.
+            AddressedOrderLimit addressed_order_limit;              
+        }
+
+        message DeletePiece {
+            bytes piece_id;
+        }
+
+        message Completed {
+            // when everything is completed
+            bytes exit_complete_signature;
+        }
+    }
+}
+```
+
+
+## Open issues (if applicable)
```

Comment:
My opinion is that all piece must be transferred regarding the durability level. The durability level should matter only to prioritize which pieces should be transferred first.

Reply:
The issue is that when we take too much time to transfer pieces people might start hard-disconnecting, because it's not worth the escrow. This could end up storage nodes not bothering with graceful exit at all.

Ideally, yes, we would transfer all pieces.

Reply:
In this case we should make sure to include the process in the design doc. We could have a configurable value `--graceful-exit.max-durability` or something that determines the cutoff where we decide we no longer care about transferring pieces on segments above a certain durability. Then we just configure it to `rs.Total` so that we transfer everything and we can lower it if we decide that takes too much time.

Reply:
Added as a future item in the overview.